server/datastore/mysql/software.go

```diff
@@ -2251,459 +2252,689 @@ func (ds *Datastore) ListCVEs(ctx context.Context, maxAge time.Duration) ([]flee
+                ua.created_at AS last_install_installed_at,
+                siua.software_installer_id AS installer_id,
+                'pending_install' AS status
+            FROM
+                upcoming_activities ua
+            INNER JOIN
+                software_install_upcoming_activities siua ON ua.id = siua.upcoming_activity_id
+            LEFT JOIN (
+                upcoming_activities ua2
+                INNER JOIN software_install_upcoming_activities siua2 ON ua2.id = siua2.upcoming_activity_id
+            ) ON ua.host_id = ua2.host_id AND
+                siua.software_installer_id = siua2.software_installer_id AND
+                ua.activity_type = ua2.activity_type AND
+                (ua2.priority < ua.priority OR ua2.created_at > ua.created_at)
+            WHERE
+                ua.host_id = :host_id AND
+                ua.activity_type = 'software_install' AND
+                ua2.id IS NULL
+        ),
+        last_software_install AS (
+            SELECT
+                hsi.execution_id AS last_install_install_uuid,
+                hsi.created_at AS last_install_installed_at,
+                hsi.software_installer_id AS installer_id,
+                hsi.status AS status
+            FROM
+                host_software_installs hsi
+            LEFT JOIN
+                host_software_installs hsi2 ON hsi.host_id = hsi2.host_id AND
+                    hsi.software_installer_id = hsi2.software_installer_id AND
+                    hsi.uninstall = hsi2.uninstall AND
+                    hsi2.removed = 0 AND
+                    hsi2.host_deleted_at IS NULL AND
+                    (hsi.created_at < hsi2.created_at OR (hsi.created_at = hsi2.created_at AND hsi.id < hsi2.id))
+            WHERE
+                hsi.host_id = :host_id AND
+                hsi.removed = 0 AND
+                hsi.uninstall = 0 AND
+                hsi.host_deleted_at IS NULL AND
+                hsi2.id IS NULL AND
+                NOT EXISTS (
+                    SELECT 1
+                    FROM
+                        upcoming_activities ua
+                    INNER JOIN
+                        software_install_upcoming_activities siua ON ua.id = siua.upcoming_activity_id
+                    WHERE
+                        ua.host_id = hsi.host_id AND
+                        siua.software_installer_id = hsi.software_installer_id AND
+                        ua.activity_type = 'software_install'
+                )
+        )
+        SELECT
+			software_installers.id AS installer_id,
+			software_titles.id AS id,
+			lsia.*
+		FROM
+			(SELECT * FROM upcoming_software_install UNION SELECT * FROM last_software_install) AS lsia
+		INNER JOIN
+			software_installers ON lsia.installer_id = software_installers.id
+		INNER JOIN
+			software_titles ON software_installers.title_id = software_titles.id
+    `
+	softwareInstallsStmt, args, err := sqlx.Named(softwareInstallsStmt, map[string]any{
```

Comment:
Now that we have more self-contained queries, can probably switch from Named to using ? directly. Same with uninstalls.